src/lib.rs

```diff
@@ -10,7 +10,61 @@ pub mod socket;
 mod topic;
 
 pub use serde_json::Value;
+use tokio::sync::broadcast;
+use tokio::time::error::Elapsed;
 
 pub use self::channel::*;
 pub use self::message::{Event, Payload, PhoenixEvent};
-pub use self::socket::{ConnectError, Socket, SocketError};
+pub use self::socket::{ConnectError, Socket, SpawnError};
+
+#[derive(thiserror::Error, Debug)]
+pub enum Error {
+    #[error(transparent)]
+    Elapsed(#[from] Elapsed),
+    #[error(transparent)]
+    Broadcast(#[from] broadcast::error::RecvError),
+    #[error(transparent)]
+    SocketError(#[from] socket::Error),
+    #[error(transparent)]
+    ChannelError(#[from] channel::Error),
+}
+impl From<socket::SpawnError> for Error {
```

Comment:
It's usually not necessary to add conversions like this because code that uses this Error type will be dealing with socket::Error/channel::Error and not the error types those are comprised of and for any exceptions the error type can be mapped.

Also this can probably be generically implemented without conflicts:
```
impl<T> From<T> for Error where T: Into<SocketError> {
    fn from(err: socket::ConnectError) -> Self {
        SocketError(err.into())
    }
}
```

Reply:
Those were added for the tests because the compiler asked for them.  I switch to the generic version.

Reply:
It works, but I have to remove the `#[from]` in `SocketError(#[from] socket::Error)` or I get conflicting implementations.

Reply:
If I do both the `socket::Error` and `channel::Error` ones, then it conflicts since it can't tell that they will only implement one:

```
impl<T> From<T> for Error where T: Into<socket::Error> {
    fn from(err: T) -> Self {
        Self::SocketError(err.into())
    }
}
impl<T> From<T> for Error where T: Into<channel::Error> {
    fn from(err: T) -> Self {
        Self::ChannelError(err.into())
    }
}
```

```
error[E0119]: conflicting implementations of trait `From<_>` for type `Error`
  --> src/lib.rs:36:1
   |
31 | impl<T> From<T> for Error where T: Into<socket::Error> {
   | ------------------------- first implementation here
...
36 | impl<T> From<T> for Error where T: Into<channel::Error> {
   | ^^^^^^^^^^^^^^^^^^^^^^^^^ conflicting implementation for `Error`
```

I'll macro it.

Reply:
The problem is that From<channel::Error> is implemented for both socket:Error and Error. If channel::Error is removed from Error and only in socket::Error, or conversely only in Error then this conflict won't exist.

Also it's unidiomatic to use Error as a suffix for error variants.

Reply:
It's also completely reasonable thing to map errors for types for whenever an error can't directly be converted into

Reply:
> Also it's unidiomatic to use `Error` as a suffix for error variants.

Fixed